insecure/corruptlibp2p/gossipsub_spammer.go

```diff
@@ -0,0 +1,118 @@
+package corruptlibp2p
+
+import (
+	"sync"
+	"testing"
+	"time"
+
+	pb "github.com/libp2p/go-libp2p-pubsub/pb"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/stretchr/testify/require"
+	corrupt "github.com/yhassanzadeh13/go-libp2p-pubsub"
+
+	"github.com/onflow/flow-go/insecure/internal"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/network/p2p"
+	p2ptest "github.com/onflow/flow-go/network/p2p/test"
+)
+
+// GossipSubRouterSpammer is a wrapper around the GossipSubRouter that allows us to
+// spam the victim with junk control messages.
+type GossipSubRouterSpammer struct {
+	atomicRouter *atomicRouter
+	SpammerNode  p2p.LibP2PNode
+}
+
+// NewGossipSubRouterSpammer is the main method tests call for spamming attacks.
+func NewGossipSubRouterSpammer(t *testing.T, sporkId flow.Identifier) *GossipSubRouterSpammer {
+	spammerNode, atomicRouter := getSpammerNode(t, sporkId)
+	return &GossipSubRouterSpammer{
+		atomicRouter: atomicRouter,
+		SpammerNode:  spammerNode,
+	}
+}
+
+// SpamIHave spams the victim with junk iHave messages.
+// msgCount is the number of iHave messages to send.
+// msgSize is the number of messageIDs to include in each iHave message.
+func (s *GossipSubRouterSpammer) SpamIHave(t *testing.T, victim p2p.LibP2PNode, ctlMessages []pb.ControlMessage) {
+	for _, ctlMessage := range ctlMessages {
+		require.True(t, s.atomicRouter.Get().SendControl(victim.Host().ID(), &ctlMessage))
+	}
+}
+
+// GenerateIHaveCtlMessages generates IHAVE control messages before they are sent so the test can prepare
+// to receive them before they are sent by the spammer.
+func (s *GossipSubRouterSpammer) GenerateIHaveCtlMessages(t *testing.T, msgCount, msgSize int) []pb.ControlMessage {
+	//var ctlMessageMap = make(map[string]pb.ControlMessage)
+	var iHaveCtlMsgs []pb.ControlMessage
+	for i := 0; i < msgCount; i++ {
+		iHaveCtlMsg := GossipSubCtrlFixture(WithIHave(msgCount, msgSize))
+
+		iHaves := iHaveCtlMsg.GetIhave()
+		require.Equal(t, msgCount, len(iHaves))
+		iHaveCtlMsgs = append(iHaveCtlMsgs, *iHaveCtlMsg)
+	}
+	return iHaveCtlMsgs
+}
+
+func (s *GossipSubRouterSpammer) WaitUntilInitialized(t *testing.T) {
+	require.Eventuallyf(t, func() bool {
+		// ensuring the spammer router has been initialized.
+		// this is needed because the router is initialized asynchronously.
+		return s.atomicRouter.Get() != nil
+	}, 1*time.Second, 100*time.Millisecond, "spammer router not set")
+	s.atomicRouter.set(s.atomicRouter.Get())
+}
+
+func getSpammerNode(t *testing.T, sporkId flow.Identifier) (p2p.LibP2PNode, *atomicRouter) {
+	router := newAtomicRouter()
+	spammerNode, _ := p2ptest.NodeFixture(
+		t,
+		sporkId,
+		t.Name(),
+		p2ptest.WithRole(flow.RoleConsensus),
```

Comment:
we should support all node roles here

Reply:
84550e52b62f0f3dcde5b179a78be6d1027b5043